README.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Node.js

This module is a part of the [Pip.Services](http://pipservices.org) polyglot microservices toolkit.

The Memcached module contains the following components: MemcachedLock and MemcachedCache for working with locks and cache on the Memcached server.

The module contains the following packages:
- **Build** - a standard factory for constructing components.
- **Cache** - cache Components in Memcached
- **Lock** - components of working with locks in Memcached

<a name="links"></a> Quick links:

* [Configuration](https://www.pipservices.org/recipies/configuration)
* [API Reference](https://pip-services3-node.github.io/pip-services3-memcached-node/globals.html)
* [Change Log](CHANGELOG.md)
* [Get Help](https://www.pipservices.org/community/help)
* [Contribute](https://www.pipservices.org/community/contribute)

## Use

Install the NPM package as
```bash
npm install pip-services-memcached-node --save
```

## Develop

For development you shall install the following prerequisites:
* Node.js 8+
* Visual Studio Code or another IDE of your choice
* Docker
* Typescript

Install dependencies:
```bash
npm install
```

Compile the code:
```bash
tsc
```

Run automated tests:
```bash
npm test
```

Generate API documentation:
```bash
./docgen.ps1
```

Before committing changes run dockerized build and test as:
```bash
./build.ps1
./test.ps1
./clear.ps1
```

## Contacts

The library is created and maintained by **Sergey Seroukhov**.


CHANGELOG.md
# <img src="https://uploads-ssl.webflow.com/5ea5d3315186cf5ec60c3ee4/5edf1c94ce4c859f2b188094_logo.svg" alt="Pip.Services Logo" width="200"> <br/> Memcached components for Node.js Changelog

## <a name="3.0.3"></a> 3.0.3 (2021-10-29) 
### Bug fixing
* Fix working with objects, added tests

## <a name="3.0.0"></a> 3.0.0 (2018-08-21) 

### Breaking changes
* Moved to a separate package

## <a name="1.2.0"></a> 1.2.0 (2018-08-10) 

### Features
* Added RedisCache
* Added RedisLock

## <a name="1.1.0"></a> 1.1.0 (2018-03-26) 

### Features
* Added PrometheusCounters and PrometheusMetricsService
* Added labels to PrometheusCounters and PrometheusMetricsService

## <a name="1.0.0"></a> 1.0.0 (2018-03-20) 

### Features
* **memcached** Added MemcachedCache
* **memcached** Added MemcachedLock
* **fluentd** Added FluentdLogger
* **elasticsearch** Added ElasticSearchLogger



test/fixtures/LockFixture.ts
let async = require('async');
let assert = require('chai').assert;

import { ILock } from 'pip-services3-components-node';

let LOCK1: string = "lock_1";
let LOCK2: string = "lock_2";
let LOCK3: string = "lock_3";

export class LockFixture {
    private _lock: ILock;

    public constructor(lock: ILock) {
        this._lock = lock;
    }

    public testTryAcquireLock(done: (err: any) => void): void {
        async.series([
            // Try to acquire lock for the first time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            },
            // Try to acquire lock for the second time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isFalse(result);
                    callback();
                });
            },
            // Release the lock
            (callback) => {
                this._lock.releaseLock(null, LOCK1, callback);
            },
            // Try to acquire lock for the third time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK1, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            }
        ], (err) => {
            this._lock.releaseLock(null, LOCK1);
            done(err);
        });
    }

    public testAcquireLock(done: (err: any) => void): void {
        async.series([
            // Acquire lock for the first time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            // Acquire lock for the second time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNotNull(err || null);
                    callback();
                });
            },
            // Release the lock
            (callback) => {
                this._lock.releaseLock(null, LOCK2, callback)
            },
            // Acquire lock for the third time
            (callback) => {
                this._lock.acquireLock(null, LOCK2, 3000, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
        ], (err) => {
            this._lock.releaseLock(null, LOCK2);
            done(err);
        });
    }

    public testReleaseLock(done: (err: any) => void): void {
        async.series([
            // Acquire lock for the first time
            (callback) => {
                this._lock.tryAcquireLock(null, LOCK3, 3000, (err, result) => {
                    assert.isNull(err || null);
                    assert.isTrue(result);
                    callback();
                });
            },
            // Release the lock for the first time
            (callback) => {
                this._lock.releaseLock(null, LOCK3, callback)
            },
            // Release the lock for the second time
            (callback) => {
                this._lock.releaseLock(null, LOCK3, callback)
            }
        ], done);
    }
    
}

test/fixtures/CacheFixture.ts
let assert = require('chai').assert;
let async = require('async');

import { ICache } from 'pip-services3-components-node';

let KEY1: string = "key1";
let KEY2: string = "key2";
let KEY3: string = "key3";
let KEY4: string = "key4";
let KEY5: string = "key5";
let KEY6: string = "key6";

let VALUE1: string = "value1";
let VALUE2 = { val: "value2" };
let VALUE3 = new Date();
let VALUE4 = [1, 2, 3, 4];
let VALUE5 = 12345;
let VALUE6 = null;

export class CacheFixture {
    private _cache: ICache = null;

    public constructor(cache: ICache) {
        this._cache = cache;
    }

    public testStoreAndRetrieve(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY2, VALUE2, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY3, VALUE3, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY4, VALUE4, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY5, VALUE5, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.store(null, KEY6, VALUE6, 5000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                setTimeout(() => {
                    callback();
                }, 500);
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE1, val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY2, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE2.val, val.val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY3, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE3.toISOString(), val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY4, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.lengthOf(val, 4)
                    assert.equal(VALUE4[0], val[0]);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY5, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNotNull(val);
                    assert.equal(VALUE5, val);

                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY6, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val);

                    callback();
                });
            }
        ], done);
    }

    public testRetrieveExpired(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                setTimeout(() => {
                    callback();
                }, 1500);
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val || null);

                    callback();
                });
            }
        ], done);
    }
    
    public testRemove(done: any): void {
        async.series([
            (callback) => {
                this._cache.store(null, KEY1, VALUE1, 1000, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.remove(null, KEY1, (err) => {
                    assert.isNull(err || null);
                    callback();
                });
            },
            (callback) => {
                this._cache.retrieve(null, KEY1, (err, val) => {
                    assert.isNull(err || null);
                    assert.isNull(val || null);

                    callback();
                });
            }
        ], done);
    }

}


test/lock/MemcachedLock.test.ts
import { ConfigParams } from 'pip-services3-commons-node';

import { MemcachedLock } from '../../src/lock/MemcachedLock';
import { LockFixture } from '../fixtures/LockFixture';

suite('MemcachedLock', ()=> {
    var _lock: MemcachedLock;
    var _fixture: LockFixture;

    setup((done) => {
        let host = process.env['MEMCACHED_SERVICE_HOST'] || 'localhost';
        let port = process.env['MEMCACHED_SERVICE_PORT'] || 11211;

        _lock = new MemcachedLock();

        let config = ConfigParams.fromTuples(
            'connection.host', host,
            'connection.port', port
        );
        _lock.configure(config);

        _fixture = new LockFixture(_lock);

        _lock.open(null, done);
    });

    teardown((done) => {
        _lock.close(null, done);
    });

    test('Try Acquire Lock', (done) => {
        _fixture.testTryAcquireLock(done);
    });

    test('Acquire Lock', (done) => {
        _fixture.testAcquireLock(done);
    });

    test('Release Lock', (done) => {
        _fixture.testReleaseLock(done);
    });

});


src/cache/MemcachedCache.ts
/** @module cache */
import { ConfigParams } from 'pip-services3-commons-node';
import { IConfigurable } from 'pip-services3-commons-node';
import { IReferences } from 'pip-services3-commons-node';
import { IReferenceable } from 'pip-services3-commons-node';
import { IOpenable } from 'pip-services3-commons-node';
import { InvalidStateException } from 'pip-services3-commons-node';
import { ConfigException } from 'pip-services3-commons-node';
import { ConnectionResolver } from 'pip-services3-components-node';
import { ICache } from 'pip-services3-components-node';

/**
 * Distributed cache that stores values in Memcaches caching service.
 * 
 * The current implementation does not support authentication.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *   - max_key_size:          maximum key length (default: 250)
 *   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
 *   - max_value:             maximum value length (default: 1048576)
 *   - pool_size:             pool size (default: 5)
 *   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - failures:              number of failures before stop retrying (default: 5)
 *   - retry:                 retry timeout in milliseconds (default: 30 sec)
 *   - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)
 * 
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 *
 * ### Example ###
 * 
 *     let cache = new MemcachedCache();
 *     cache.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 11211
 *     ));
 * 
 *     cache.open("123", (err) => {
 *       ...
 *     });
 * 
 *     cache.store("123", "key1", "ABC", (err) => {
 *          cache.store("123", "key1", (err, value) => {
 *              // Result: "ABC"
 *          });
 *     });
 */
export class MemcachedCache implements ICache, IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();

    private _maxKeySize: number = 250;
    private _maxExpiration: number = 2592000;
    private _maxValue: number = 1048576;
    private _poolSize: number = 5;
    private _reconnect: number = 10000;
    private _timeout: number = 5000;
    private _retries: number = 5;
    private _failures: number = 5;
    private _retry: number = 30000;
    private _remove: boolean = false;
    private _idle: number = 5000;

    private _client: any = null;

    /**
     * Creates a new instance of this cache.
     */
    public constructor() { }

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        this._connectionResolver.configure(config);

        this._maxKeySize = config.getAsIntegerWithDefault('options.max_key_size', this._maxKeySize);
        this._maxExpiration = config.getAsLongWithDefault('options.max_expiration', this._maxExpiration);
        this._maxValue = config.getAsLongWithDefault('options.max_value', this._maxValue);
        this._poolSize = config.getAsIntegerWithDefault('options.pool_size', this._poolSize);
        this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
        this._failures = config.getAsIntegerWithDefault('options.failures', this._failures);
        this._retry = config.getAsIntegerWithDefault('options.retry', this._retry);
        this._remove = config.getAsBooleanWithDefault('options.remove', this._remove);
        this._idle = config.getAsIntegerWithDefault('options.idle', this._idle);
    }

    /**
     * Sets references to dependent components.
     * 
     * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
    }

    /**
     * Checks if the component is opened.
     * 
     * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
     * Opens the component.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public open(correlationId: string, callback: (err: any) => void): void {
        this._connectionResolver.resolveAll(correlationId, (err, connections) => {
            if (err == null && connections.length == 0)
                err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');

            if (err != null) {
                callback(err);
                return;
            }

            let servers: string[] = [];
            for (let connection of connections) {
                let host = connection.getHost();
                let port = connection.getPort() || 11211;
                servers.push(host + ':' + port);
            }

            let options = {
                maxKeySize: this._maxKeySize,
                maxExpiration: this._maxExpiration,
                maxValue: this._maxValue,
                poolSize: this._poolSize,
                reconnect: this._reconnect,
                timeout: this._timeout,
                retries: this._retries,
                failures: this._failures,
                retry: this._retry,
                remove: this._remove,
                idle: this._idle
            };

            let Memcached = require('memcached');
            this._client = new Memcached(servers, options);

            if (callback) callback(null);
        });
    }

    /**
     * Closes component and frees used resources.
     * 
     * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public close(correlationId: string, callback: (err: any) => void): void {
        this._client = null;
        if (callback) callback(null);
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpen()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }

        return true;
    }

    /**
     * Retrieves cached value from the cache using its key.
     * If value is missing in the cache or expired it returns null.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param callback          callback function that receives cached value or error.
     */
    public retrieve(correlationId: string, key: string,
        callback: (err: any, value: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.get(key, (err, item) => {
            if (err) {
                callback(err, null);
                return;
            }
            callback(err, item ? JSON.parse(item) : item);
        });
    }

    /**
     * Stores value in the cache with expiration time.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param value             a value to store.
     * @param timeout           expiration timeout in milliseconds.
     * @param callback          (optional) callback function that receives an error or null for success
     */
    public store(correlationId: string, key: string, value: any, timeout: number,
        callback: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        let timeoutInSec = timeout / 1000;
        this._client.set(key, JSON.stringify(value), timeoutInSec, callback);
    }

    /**
     * Removes a value from the cache by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique value key.
     * @param callback          (optional) callback function that receives an error or null for success
     */
    public remove(correlationId: string, key: string,
        callback: (err: any) => void) {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.del(key, callback);
    }

}

src/lock/MemcachedLock.ts
/** @module lock */
import { ConfigParams } from 'pip-services3-commons-node';
import { IConfigurable } from 'pip-services3-commons-node';
import { IReferences } from 'pip-services3-commons-node';
import { IReferenceable } from 'pip-services3-commons-node';
import { IOpenable } from 'pip-services3-commons-node';
import { InvalidStateException } from 'pip-services3-commons-node';
import { ConfigException } from 'pip-services3-commons-node';
import { ConnectionResolver } from 'pip-services3-components-node';
import { Lock } from 'pip-services3-components-node';

/**
 * Distributed lock that implemented based on Memcaches caching service.
 * 
 * The current implementation does not support authentication.
 * 
 * ### Configuration parameters ###
 * 
 * - connection(s):           
 *   - discovery_key:         (optional) a key to retrieve the connection from [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]]
 *   - host:                  host name or IP address
 *   - port:                  port number
 *   - uri:                   resource URI or connection string with all parameters in it
 * - options:
 *   - retry_timeout:         timeout in milliseconds to retry lock acquisition. (Default: 100)
 *   - max_size:              maximum number of values stored in this cache (default: 1000)        
 *   - max_key_size:          maximum key length (default: 250)
 *   - max_expiration:        maximum expiration duration in milliseconds (default: 2592000)
 *   - max_value:             maximum value length (default: 1048576)
 *   - pool_size:             pool size (default: 5)
 *   - reconnect:             reconnection timeout in milliseconds (default: 10 sec)
 *   - retries:               number of retries (default: 3)
 *   - timeout:               default caching timeout in milliseconds (default: 1 minute)
 *   - failures:              number of failures before stop retrying (default: 5)
 *   - retry:                 retry timeout in milliseconds (default: 30 sec)
 *   - idle:                  idle timeout before disconnect in milliseconds (default: 5 sec)
 * 
 * ### References ###
 * 
 * - <code>\*:discovery:\*:\*:1.0</code>        (optional) [[https://pip-services3-node.github.io/pip-services3-components-node/interfaces/connect.idiscovery.html IDiscovery]] services to resolve connection
 *
 * ### Example ###
 * 
 *     let lock = new MemcachedLock();
 *     lock.configure(ConfigParams.fromTuples(
 *       "host", "localhost",
 *       "port", 11211
 *     ));
 * 
 *     lock.open("123", (err) => {
 *       ...
 *     });
 * 
 *     lock.acquire("123", "key1", (err) => {
 *          if (err == null) {
 *              try {
 *                // Processing...
 *              } finally {
 *                 lock.releaseLock("123", "key1", (err) => {
 *                    // Continue...
 *                 });
 *              }
 *          }
 *     });
 */
export class MemcachedLock extends Lock implements IConfigurable, IReferenceable, IOpenable {
    private _connectionResolver: ConnectionResolver = new ConnectionResolver();
    
    private _maxKeySize: number = 250;
    private _maxExpiration: number = 2592000;
    private _maxValue: number = 1048576;
    private _poolSize: number = 5;
    private _reconnect: number = 10000;
    private _timeout: number = 5000;
    private _retries: number = 5;
    private _failures: number = 5;
    private _retry: number = 30000;
    private _remove: boolean = false;
    private _idle: number = 5000;

    private _client: any = null;

    /**
     * Configures component by passing configuration parameters.
     * 
     * @param config    configuration parameters to be set.
     */
    public configure(config: ConfigParams): void {
        super.configure(config);

        this._connectionResolver.configure(config);

        this._maxKeySize = config.getAsIntegerWithDefault('options.max_key_size', this._maxKeySize);
        this._maxExpiration = config.getAsLongWithDefault('options.max_expiration', this._maxExpiration);
        this._maxValue = config.getAsLongWithDefault('options.max_value', this._maxValue);
        this._poolSize = config.getAsIntegerWithDefault('options.pool_size', this._poolSize);
        this._reconnect = config.getAsIntegerWithDefault('options.reconnect', this._reconnect);
        this._timeout = config.getAsIntegerWithDefault('options.timeout', this._timeout);
        this._retries = config.getAsIntegerWithDefault('options.retries', this._retries);
        this._failures = config.getAsIntegerWithDefault('options.failures', this._failures);
        this._retry = config.getAsIntegerWithDefault('options.retry', this._retry);
        this._remove = config.getAsBooleanWithDefault('options.remove', this._remove);
        this._idle = config.getAsIntegerWithDefault('options.idle', this._idle);
    }

    /**
	 * Sets references to dependent components.
	 * 
	 * @param references 	references to locate the component dependencies. 
     */
    public setReferences(references: IReferences): void {
        this._connectionResolver.setReferences(references);
    }

    /**
	 * Checks if the component is opened.
	 * 
	 * @returns true if the component has been opened and false otherwise.
     */
    public isOpen(): boolean {
        return this._client;
    }

    /**
	 * Opens the component.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public open(correlationId: string, callback: (err: any) => void): void {
        this._connectionResolver.resolveAll(correlationId, (err, connections) => {
            if (err == null && connections.length == 0)
                err = new ConfigException(correlationId, 'NO_CONNECTION', 'Connection is not configured');

            if (err != null) {
                 callback(err);
                 return;
            } 

            let servers: string[] = [];
            for (let connection of connections) {
                let host = connection.getHost();
                let port = connection.getPort() || 11211;
                servers.push(host + ':' + port);
            }

            let options = {
                maxKeySize: this._maxKeySize,
                maxExpiration: this._maxExpiration,
                maxValue: this._maxValue,
                poolSize: this._poolSize,
                reconnect: this._reconnect,
                timeout: this._timeout,
                retries: this._retries,
                failures: this._failures,
                retry: this._retry,
                remove: this._remove,
                idle: this._idle
            };

            let Memcached = require('memcached');
            this._client = new Memcached(servers, options);

            if (callback) callback(null);
        });
    }

    /**
	 * Closes component and frees used resources.
	 * 
	 * @param correlationId 	(optional) transaction id to trace execution through call chain.
     * @param callback 			callback function that receives error or null no errors occured.
     */
    public close(correlationId: string, callback: (err: any) => void): void {
        this._client = null;
        if (callback) callback(null);
    }

    private checkOpened(correlationId: string, callback: any): boolean {
        if (!this.isOpen()) {
            let err = new InvalidStateException(correlationId, 'NOT_OPENED', 'Connection is not opened');
            callback(err, null);
            return false;
        }
        
        return true;
    }
    
    /**
     * Makes a single attempt to acquire a lock by its key.
     * It returns immediately a positive or negative result.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to acquire.
     * @param ttl               a lock timeout (time to live) in milliseconds.
     * @param callback          callback function that receives a lock result or error.
     */
    public tryAcquireLock(correlationId: string, key: string, ttl: number,
        callback: (err: any, result: boolean) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        let lifetimeInSec = ttl / 1000;
        this._client.add(key, 'lock', lifetimeInSec, (err) => {
            if (err && err.message && err.message.indexOf('not stored') >= 0)
                callback(null, false);
            else callback(err, err == null);
        });
    }

    /**
     * Releases prevously acquired lock by its key.
     * 
     * @param correlationId     (optional) transaction id to trace execution through call chain.
     * @param key               a unique lock key to release.
     * @param callback          callback function that receives error or null for success.
     */
    public releaseLock(correlationId: string, key: string,
        callback?: (err: any) => void): void {
        if (!this.checkOpened(correlationId, callback)) return;

        this._client.del(key, callback);
    }
}